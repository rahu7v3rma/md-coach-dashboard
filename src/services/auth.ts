const AUTH_KEY = 'md.coach.token';

export const AUTHORIZATION_HEADER_NAME = 'X-Authorization';

export const setAuthToken = (token: string) => {
    sessionStorage.setItem(AUTH_KEY, token);
};

const getAuthToken = () => {
    return sessionStorage.getItem(AUTH_KEY);
};

export const resetAuthToken = () => {
    sessionStorage.removeItem(AUTH_KEY);
};

export const getAuthorizationHeaderValue = () => {
    const token = getAuthToken();
    return `Token ${token}`;
};

export const hasAuthToken = () => {
    return !!getAuthToken();
};
