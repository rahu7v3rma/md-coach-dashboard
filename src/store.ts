import { configureStore } from '@reduxjs/toolkit';

import clientReducer from './reducers/client';
import groupReducer from './reducers/group';
import notificationReducer from './reducers/notification';
import userReducer from './reducers/user';

// define store options as a variable so we can export it for tests to use
const storeOptions: any = {
    reducer: {
        user: userReducer,
        client: clientReducer,
        notification: notificationReducer,
        group: groupReducer
    }
};

const store = configureStore(storeOptions);

export default store;

// infer the `RootState` type from the store
export type RootState = ReturnType<typeof store.getState>;

// infer the `AppDispatch` type from the store so thunk actions can return promises
export type AppDispatch = typeof store.dispatch;

export const test = {
    storeOptions
};
