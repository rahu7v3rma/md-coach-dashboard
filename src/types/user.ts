export type UserProfile = {
    image: string | null;
};

export type UserChatProfile = {
    apiKey: string;
    token: string;
    userId: string;
};

export type UserInfo = {
    email: string;
    first_name: string;
    last_name: string;
    image: string | null;
    diabetes_type: string;
    date_of_birth: string;
    contact_number: string;
    chat_id: string;
    onboarding_answers?: OnboardingAnswer[];
};

export type Last7Days = {
    fast: { duration_minutes_average?: number };
    exercise: { duration_minutes_sum?: number };
    glucose: { amount_avg?: number };
};

export type LastLog = {
    fast: Fast;
    weight: weight;
    exercise: Exercise;
    drink: Drink;
    glucose: Glucose;
};

export type Fast = {
    duration_minutes: number;
};

export type weight = {
    amount: number;
    unit: string;
    previous_amount?: number;
};

export type Exercise = {
    type: string;
    duration_minutes: number;
};

export type Drink = {
    amount: number;
};

export type Glucose = {
    amount: number;
    unit: string;
    previous_amount?: number;
};

export type Profile = {
    diabetes_type: string;
};

export type UserClient = {
    activity_type: string;
    duration: number;
    engagement: string;
    last_fasting: number;
    second_last_fasting: number;
    hydration: number;
    hydration_unit: string;
    id: number;
    last_week_avg_fbg: number;
    name: string;
    type: string;
    group: string;
    weekly_total: number;
    last_weight: number;
    second_last_weight: number;
    weight_units: string;
    fbg: number | string;
    second_last_fbg: number | string;
    fbg_unit: string;
    weight_unit: string;
};

export type member = {
    id: number;
    first_name: string;
    last_name: string;
};

export type UserGroup = {
    id: number;
    members: member[];
    name: string;
    group_type: string;
    program: string;
    attendance_sheet_url: string;
};

export type OnboardingAnswer = {
    question: string;
    answer: string;
};
