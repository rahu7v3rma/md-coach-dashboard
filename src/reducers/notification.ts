import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from 'src/store';
import { NotificationListResponse } from 'src/types/notification';

interface NotificationState {
    loading: boolean;
    notifications: NotificationListResponse;
    unreadCount: number;
}

const userAdapter = createEntityAdapter();

const initialState = userAdapter.getInitialState({
    loading: false,
    notifications: {
        list: [],
        hasPrevious: false,
        hasNext: false,
        totalCount: 0,
        nextPageNumber: 1,
        previousPageNumber: 0
    },
    unreadCount: 0
} as NotificationState);

export const getNotification = createAsyncThunk(
    'user/notifications',
    async (
        options: {
            page: number;
            limit: number;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.getNotifction(options.page, options.limit);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const readNotification = createAsyncThunk(
    'user/readNotifications',
    async (
        options: {
            id: number;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.readNotification(options.id);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

const notificationSlice = createSlice({
    name: 'notification',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(getNotification.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getNotification.fulfilled, (state, action) => {
            state.loading = false;
            if (state.notifications.hasPrevious) {
                const oldData = state.notifications.list;
                action.payload.list = [...oldData, ...action.payload.list];
            }
            state.notifications.list = action.payload.list;
            state.unreadCount = action.payload.total_unread;
            state.loading = false;
        });
        builder.addCase(getNotification.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(readNotification.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(readNotification.fulfilled, (state, action) => {
            state.loading = false;
            state.notifications.list = state.notifications.list.map((item) => {
                if (item.id === action.meta.arg.id) {
                    item.read_flag = true;
                    state.unreadCount = state.unreadCount - 1;
                }
                return item;
            });
        });
        builder.addCase(readNotification.rejected, (state) => {
            state.loading = false;
        });
    }
});

export default notificationSlice.reducer;

interface NotificationSelectorsType {
    loading: boolean;
    notifications: NotificationListResponse;
    unreadCount: number;
}

export const NotificationSelectors = (): NotificationSelectorsType => {
    const loading = useSelector(
        (state: RootState) => state.notification.loading
    );

    const notifications = useSelector(
        (state: RootState) => state.notification.notifications
    );

    const unreadCount = useSelector(
        (state: RootState) => state.notification.unreadCount
    );
    return {
        loading,
        notifications,
        unreadCount
    };
};
