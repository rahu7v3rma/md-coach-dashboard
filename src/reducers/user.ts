import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from '../store';
import { UserChatProfile, UserProfile } from '../types/user';
import { resetAuthToken } from 'src/services/auth';

const userAdapter = createEntityAdapter();

interface UserState {
    loading: boolean;
    userProfile: UserProfile | null;
    chatProfile: UserChatProfile | null;
    isSideBarOpen: boolean;
}

const initialState = userAdapter.getInitialState({
    loading: false,
    userProfile: null,
    chatProfile: null,
    isSideBarOpen: false
} as UserState);

/**
 * Login api - Use promise to get the result of login, since no need to set the login result/error in global state
 */
export const login = createAsyncThunk(
    'user/login',
    async (
        options: { email: string; password: string },
        { rejectWithValue }
    ) => {
        try {
            return await api.login(options.email, options.password);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

/**
 * Reset password api - Use promise to get the result of reset password, since no need to set the reset password result/error in global state
 */
export const resetPassword = createAsyncThunk(
    'user/reset/request',
    async (options: { email: string; client: string }, { rejectWithValue }) => {
        try {
            return await api.resetPassword(options.email, options.client);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const resetPasswordVerify = createAsyncThunk(
    'user/resetPasswordVerify',
    async (options: { code: string }, { rejectWithValue }) => {
        try {
            return await api.resetPasswordVerify(options.code);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const resetPasswordChange = createAsyncThunk(
    'user/resetPasswordChange',
    async (
        options: {
            code: string;
            password: string;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.resetPasswordChange(
                options.code,
                options.password
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const refreshProfileSession = createAsyncThunk(
    'user/refreshProfileSession',
    async (_options: Record<string, string>, { rejectWithValue }) => {
        try {
            return await api.refreshProfileSession();
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getProfile = createAsyncThunk(
    'user/getProfile',
    async (_options: {}, { rejectWithValue }) => {
        try {
            return await api.getProfile();
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const updateProfile = createAsyncThunk(
    'user/updateProfile',
    async (options: { image: string }, { rejectWithValue }) => {
        try {
            return await api.updateProfile(options.image);
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const logoutUser = createAsyncThunk(
    'user/logoutUser',
    async (options: {}, { rejectWithValue }) => {
        try {
            return await api.logoutUser();
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

const userSlice = createSlice({
    name: 'user',
    initialState,
    reducers: {
        toggleSidebar: (state, action) => {
            state.isSideBarOpen = action.payload.isOpen;
        }
    },
    extraReducers: (builder) => {
        builder.addCase(login.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(login.fulfilled, (state) => {
            state.loading = false;
        });
        builder.addCase(login.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPassword.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(resetPassword.fulfilled, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPassword.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPasswordVerify.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(resetPasswordVerify.fulfilled, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPasswordVerify.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPasswordChange.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(resetPasswordChange.fulfilled, (state) => {
            state.loading = false;
        });
        builder.addCase(resetPasswordChange.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(refreshProfileSession.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(refreshProfileSession.fulfilled, (state, action) => {
            state.loading = false;
            state.chatProfile = action.payload;
        });
        builder.addCase(refreshProfileSession.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getProfile.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getProfile.fulfilled, (state, action) => {
            state.loading = false;
            state.userProfile = action.payload;
        });
        builder.addCase(getProfile.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(updateProfile.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(updateProfile.fulfilled, (state, action) => {
            state.loading = false;

            if (state.userProfile) {
                state.userProfile.image = action.meta.arg.image;
            }
        });
        builder.addCase(updateProfile.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(logoutUser.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(logoutUser.fulfilled, (state) => {
            state.loading = false;
            resetAuthToken();
            state = initialState;
        });
        builder.addCase(logoutUser.rejected, (state) => {
            state.loading = false;
        });
    }
});

export const { toggleSidebar } = userSlice.actions;

export default userSlice.reducer;

interface UserSelectorsType {
    loading: boolean;
    userProfile: UserProfile | null;
    chatProfile: UserChatProfile | null;
    isSideBarOpen: boolean;
}

export const UserSelectors = (): UserSelectorsType => {
    const loading = useSelector((state: RootState) => state.user.loading);

    const userProfile = useSelector(
        (state: RootState) => state.user.userProfile
    );

    const chatProfile = useSelector(
        (state: RootState) => state.user.chatProfile
    );

    const isSideBarOpen = useSelector(
        (state: RootState) => state.user.isSideBarOpen
    );

    return {
        loading,
        userProfile,
        chatProfile,
        isSideBarOpen
    };
};
