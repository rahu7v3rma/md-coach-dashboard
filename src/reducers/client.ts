import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import moment from 'moment';
import { SortOrder } from 'react-data-table-component';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from '../store';
import { Colors } from '../utils/colors';
import { ClientProfiles } from 'src/types/client';
import { logBooks } from 'src/types/log';
import { Note, Notes } from 'src/types/note';
import { UserClient, UserGroup } from 'src/types/user';

const clientAdapter = createEntityAdapter();

interface ClientState {
    loading: boolean;
    clients: {
        list: Array<UserClient>;
        searchInput: string;
        hasPrevious: boolean;
        hasNext: boolean;
        totalCount: number;
        nextPageNumber: number;
        previousPageNumber: number;
        sortBy: number;
        order: SortOrder;
    };
    profiles: ClientProfiles;
    logBook: logBooks;
    notes: Notes;
    groups: {
        list: Array<UserGroup>;
        searchInput: string;
        hasPrevious: boolean;
        hasNext: boolean;
        totalCount: number;
        nextPageNumber: number;
        previousPageNumber: number;
        sortBy: number;
        order: SortOrder;
    };
    analytics: any;
}

const initialState = clientAdapter.getInitialState({
    loading: false,
    profiles: {},
    logBook: {},
    notes: {},
    clients: {
        list: [],
        searchInput: '',
        hasPrevious: false,
        hasNext: false,
        totalCount: 0,
        nextPageNumber: 1,
        previousPageNumber: 0,
        sortBy: 0,
        order: 'asc'
    },
    groups: {
        list: [],
        searchInput: '',
        hasPrevious: false,
        hasNext: false,
        totalCount: 4,
        nextPageNumber: 1,
        previousPageNumber: 0,
        sortBy: 0,
        order: 'asc'
    },
    analytics: []
} as ClientState);

export const getUserInfo = createAsyncThunk(
    'client/getUserInfo',
    async (id: string, { rejectWithValue }) => {
        try {
            const userInfo = await api.getUserInfo(id);
            return {
                ...userInfo,
                id
            };
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getUserAnalytics = createAsyncThunk(
    'client/getUserAnalytics',
    async (
        { id, startDate, endDate }: Record<string, string | number>,
        { rejectWithValue }
    ) => {
        try {
            return await api.getUserAnalytics(
                id as string | number,
                startDate as string,
                endDate as string
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getClients = createAsyncThunk(
    'client/customers',
    async (
        { page, limit, search, sortBy, order }: Record<string, string | number>,
        { rejectWithValue }
    ) => {
        try {
            return await api.getClients(
                page as number,
                limit as number,
                search as string,
                sortBy as number,
                order as SortOrder
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const getLogBook = createAsyncThunk(
    'user/getLogBook',
    async (
        options: {
            page: number;
            limit: number;
            user_id: string;
            category: string;
        },
        { rejectWithValue }
    ) => {
        try {
            return await api.getLogBook(
                options.page,
                options.limit,
                options.user_id,
                options.category
            );
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

export const onSearchChange = createAsyncThunk(
    'onSearchChange',
    async (search: string) => {
        return search;
    }
);

export const getNotes = createAsyncThunk(
    'client/getNotes',
    async (options: { page: number; limit: number; clientId: string }) => {
        return await api.getNotes(
            options.page,
            options.limit,
            options.clientId
        );
    }
);

export const addNote = createAsyncThunk(
    'client/addNote',
    async (options: {
        id: number;
        coach_id?: string;
        client_id: string;
        content: string;
        created_at: Date;
    }): Promise<Note> => {
        return await api.addNote(+options.client_id, options.content);
    }
);

export const editNote = createAsyncThunk(
    'client/editNote',
    async (options: {
        id: number;
        coach_id?: string;
        client_id: string;
        content: string;
        created_at: Date;
    }): Promise<Note> => {
        return await api.updateNote(
            options.id,
            +options.client_id,
            options.content
        );
    }
);

export const deleteNote = createAsyncThunk(
    'client/deleteNote',
    async (options: { id: number; clientId: string }) => {
        return await api.deleteNote(options.id);
    }
);

const clientSlice: any = createSlice({
    name: 'client',
    initialState,
    reducers: {
        sortOrder: (state, action) => {
            state.clients.sortBy = action.payload.sortBy;
            state.clients.order = action.payload.order;
        },
        sortGroupOrder: (state, action) => {
            state.groups.sortBy = action.payload.sortBy;
            state.groups.order = action.payload.order;
        }
    },
    extraReducers: (builder) => {
        builder.addCase(getLogBook.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getLogBook.fulfilled, (state, action) => {
            state.loading = false;
            const userID = action.meta.arg.user_id;
            if (action.payload.has_previous) {
                const old_list = state.logBook[userID].list;
                const appended_list = [...old_list, ...action.payload.list];
                action.payload.list = appended_list;
                state.logBook[userID] = action.payload;
            } else {
                state.logBook[userID] = action.payload;
            }
        });
        builder.addCase(getLogBook.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getUserInfo.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getUserInfo.fulfilled, (state, action) => {
            state.loading = false;

            const profiles: ClientProfiles = { ...state.profiles };
            profiles[action.payload.id] = action.payload;

            state.profiles = profiles;
        });
        builder.addCase(getUserInfo.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getClients.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getClients.fulfilled, (state, action) => {
            state.loading = false;
            state.clients.hasPrevious = action.payload.has_previous;
            state.clients.hasNext = action.payload.has_next;
            state.clients.nextPageNumber = action.payload.next_page_number;
            state.clients.totalCount = action.payload.count;
            state.clients.list = action.payload.list;
            state.clients.sortBy = action.meta.arg.sortBy as number;
            state.clients.order = action.meta.arg.order as SortOrder;
        });
        builder.addCase(getClients.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(onSearchChange.fulfilled, (state, action) => {
            state.clients.searchInput = action.payload;
        });
        builder.addCase(getNotes.fulfilled, (state, action) => {
            const userID = action.meta.arg.clientId;
            if (action.payload.has_previous) {
                const old_list = state.notes[userID].list;
                const appended_list = [...old_list, ...action.payload.list];
                action.payload.list = appended_list;
                state.notes[userID] = action.payload;
            } else {
                state.notes[userID] = action.payload;
            }
        });
        builder.addCase(addNote.fulfilled, (state, action) => {
            state.notes[action.meta.arg.client_id].list.push(action.payload);
        });
        builder.addCase(editNote.fulfilled, (state, action) => {
            let previous_notes = {
                ...state.notes[action.meta.arg.client_id]
            };
            let edited_note = previous_notes.list.filter(
                (note) => note.id === action.meta.arg.id
            );
            if (edited_note.length) {
                edited_note[0].content = action.payload.content;
            }
            state.notes[action.meta.arg.client_id] = previous_notes;
        });
        builder.addCase(deleteNote.fulfilled, (state, action) => {
            let previous_notes = {
                ...state.notes[action.meta.arg.clientId]
            };
            previous_notes.list = previous_notes.list.filter(
                (note) => note.id !== action.meta.arg.id
            );
            state.notes[action.meta.arg.clientId] = previous_notes;
        });
        builder.addCase(getUserAnalytics.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(getUserAnalytics.rejected, (state) => {
            state.loading = false;
        });
        builder.addCase(getUserAnalytics.fulfilled, (state, action: any) => {
            const UserFastData: any =
                Object.keys(action.payload.UserFast).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserFast[key]?.length > 0
                            ? action.payload.UserFast[key][0]
                                  ?.duration_minutes ?? 0
                            : 0,
                    unit: 'min'
                })) ?? [];
            const UserGlucoseData: any =
                Object.keys(action.payload.UserGlucose).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserGlucose[key]?.length > 0
                            ? action.payload.UserGlucose[key][0]?.amount ?? 0
                            : 0,
                    unit:
                        action.payload.UserGlucose[key]?.length > 0
                            ? action.payload.UserGlucose[key][0]?.unit ?? ''
                            : ''
                })) ?? [];
            const UserWeightData: any =
                Object.keys(action.payload.UserWeight).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserWeight[key]?.length > 0
                            ? Math.round(
                                  action.payload.UserWeight[key][0]?.amount
                              ) ?? 0
                            : 0,
                    unit:
                        action.payload.UserWeight[key]?.length > 0
                            ? action.payload.UserWeight[key][0]?.unit ?? ''
                            : ''
                })) ?? [];
            const UserInsulinData: any =
                Object.keys(action.payload.UserInsulin).map((key: string) => ({
                    x: moment(key).format('DD/MM/YY'),
                    y:
                        action.payload.UserInsulin[key]?.length > 0
                            ? action.payload.UserInsulin[key][0]?.units ?? 0
                            : 0,
                    unit:
                        action.payload.UserInsulin[key]?.length > 0
                            ? action.payload.UserInsulin[key][0]
                                  ?.injection_type ?? ''
                            : ''
                })) ?? [];

            state.analytics = [
                {
                    title: 'Glucose',
                    color: Colors.extra.darkTerraCotta,
                    chart: {
                        type: 'line' as 'line',
                        data: UserGlucoseData
                    }
                },
                {
                    title: 'Weight (kg)',
                    color: Colors.extra.skyBlue2,
                    chart: {
                        type: 'line' as 'line',
                        data: UserWeightData
                    }
                },
                {
                    title: 'Fasting',
                    color: Colors.extra.lightPink,
                    chart: {
                        type: 'bar' as 'bar',
                        data: UserFastData
                    }
                },
                {
                    title: 'Insulin',
                    color: Colors.extra.lightGreen2,
                    chart: {
                        type: 'bar' as 'bar',
                        data: UserInsulinData
                    }
                }
            ];
        });
    }
});

export default clientSlice.reducer;

interface ClientSelectorsType {
    loading: boolean;
    logBook: any;
    notes: Notes;
    profiles: ClientProfiles;
    clients: Array<UserClient>;
    searchInput: string;
    hasPrevious: boolean;
    hasNext: boolean;
    totalCount: number;
    totalGroupCount: number;
    nextPageNumber: number;
    previousPageNumber: number;
    sortBy: number;
    order: string;
    groups: Array<UserGroup>;
    groupSortBy: number;
    groupOrder: string;
    analytics: any;
}

export const { sortOrder, sortGroupOrder } = clientSlice.actions;

export const ClientSelectors = (): ClientSelectorsType => {
    const loading = useSelector((state: RootState) => state.client.loading);
    const clients = useSelector(
        (state: RootState) => state.client.clients.list
    );
    const groups = useSelector((state: RootState) => state.client.groups.list);
    const hasPrevious = useSelector(
        (state: RootState) => state.client.clients.hasPrevious
    );
    const hasNext = useSelector(
        (state: RootState) => state.client.clients.hasNext
    );
    const totalGroupCount = useSelector(
        (state: RootState) => state.client.groups.totalCount
    );
    const totalCount = useSelector(
        (state: RootState) => state.client.clients.totalCount
    );
    const searchInput = useSelector(
        (state: RootState) => state.client.clients.searchInput
    );
    const nextPageNumber = useSelector(
        (state: RootState) => state.client.clients.nextPageNumber
    );
    const previousPageNumber = useSelector(
        (state: RootState) => state.client.clients.previousPageNumber
    );

    const sortBy = useSelector(
        (state: RootState) => state.client.clients.sortBy
    );

    const order = useSelector((state: RootState) => state.client.clients.order);

    const groupSortBy = useSelector(
        (state: RootState) => state.client.groups.sortBy
    );

    const groupOrder = useSelector(
        (state: RootState) => state.client.groups.order
    );

    const profiles = useSelector((state: RootState) => state.client.profiles);

    const logBook = useSelector((state: RootState) => state.client.logBook);

    const notes = useSelector((state: RootState) => state.client.notes);

    const analytics = useSelector((state: RootState) => state.client.analytics);

    return {
        loading,
        logBook,
        notes,
        profiles,
        clients,
        searchInput,
        hasPrevious,
        hasNext,
        totalCount,
        nextPageNumber,
        previousPageNumber,
        sortBy,
        order,
        groups,
        totalGroupCount,
        groupSortBy,
        groupOrder,
        analytics
    };
};
