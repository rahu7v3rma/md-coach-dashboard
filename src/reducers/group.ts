import {
    createAsyncThunk,
    createEntityAdapter,
    createSlice
} from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import * as api from '../services/api';
import { RootState } from 'src/store';
import { UserGroup } from 'src/types/user';

interface GroupState {
    loading: boolean;
    groups: UserGroup[];
}

const groupAdapter = createEntityAdapter();

const initialState = groupAdapter.getInitialState({
    loading: false
} as GroupState);

export const fetchGroupsList = createAsyncThunk(
    'chat/groups',
    async (options: {}, { rejectWithValue }) => {
        try {
            return await api.getGroupsList();
        } catch (err) {
            return rejectWithValue(err);
        }
    }
);

const groupSlice = createSlice({
    name: 'group',
    initialState,
    reducers: {},
    extraReducers: (builder) => {
        builder.addCase(fetchGroupsList.pending, (state) => {
            state.loading = true;
        });
        builder.addCase(fetchGroupsList.fulfilled, (state, action) => {
            state.loading = false;
            state.groups = action.payload;
        });
        builder.addCase(fetchGroupsList.rejected, (state) => {
            state.loading = false;
        });
    }
});

export default groupSlice.reducer;

interface GroupSelectorsType {
    loading: boolean;
    groups: UserGroup[];
}

export const GroupSelectors = (): GroupSelectorsType => {
    const loading = useSelector((state: RootState) => state.group.loading);
    const groups = useSelector((state: RootState) => state.group.groups);
    return {
        loading,
        groups
    };
};
