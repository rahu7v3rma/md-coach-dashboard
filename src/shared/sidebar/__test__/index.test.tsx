import React from 'react';
import { Provider } from 'react-redux';
import * as router from 'react-router';
import { BrowserRouter } from 'react-router-dom';
import { act, create } from 'react-test-renderer';

import Sidebar from '..';
import { logoutUser } from '../../../reducers/user';
import * as auth from '../../../services/auth';
import store from '../../../store';

const navigate = jest.fn();
const mockRequestPermissions = jest.fn();

jest.mock('../../../hooks', () => ({
    useAppDispatch: () => {
        return jest.fn().mockImplementation(() => {
            return {
                unwrap: jest.fn().mockImplementation(() => {
                    return Promise.resolve({});
                })
            };
        });
    },
    useNotificationsPermissions: jest
        .fn()
        .mockImplementationOnce(() => {
            return {
                hasPermissions: true
            };
        })
        .mockImplementation(() => {
            return {
                hasPermissions: false
            };
        })
}));

jest.mock('../../../services/notification', () => ({
    requestPermissions: jest.fn().mockImplementation(() => {
        mockRequestPermissions();
    })
}));

const mockApiResponse = () => {
    (fetch as any).mockResponse(() => {
        return Promise.resolve(
            JSON.stringify({
                id_token: '',
                refresh_token: '',
                access_token: '',
                data: {}
            })
        );
    });
};

describe('SideBar', () => {
    beforeEach(() => {
        jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
        jest.clearAllMocks();
    });

    it('does not display the notification notice when permissions are present', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        // notification permission mock for permission
        let notificationNotice = tree.root.findAllByProps({
            children:
                'Notifications are disabled. Click here to grant required permissions'
        });
        expect(notificationNotice.length).toBe(0);
    });

    it('SideBar snapshot', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        expect(tree).toMatchSnapshot();
    });

    it('SideBar renders without errors', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        expect(tree.root).toBeTruthy();
    });

    it('renders the logo and menu items', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        let logoIcon = tree.root.findByProps({
            alt: 'MD Icon'
        }).props;
        expect(logoIcon).toBeTruthy();
        let sideBarClientItem = tree.root.findByProps({
            name: 'My Clients'
        }).props;
        expect(sideBarClientItem).toBeTruthy();
        let sideBarChatItem = tree.root.findByProps({
            name: 'Chat'
        }).props;
        expect(sideBarChatItem).toBeTruthy();
    });

    it('displays a notification notice when permissions are missing', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        // notification permission mock for missing permission
        let notificationNotice = tree.root.findByProps({
            children:
                'Notifications are disabled. Click here to grant required permissions'
        }).props;
        expect(notificationNotice).toBeTruthy();
    });

    it('requests permissions when the notice is clicked', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        // notification permission mock for missing permission
        let notificationNotice = tree.root.findByProps({
            children:
                'Notifications are disabled. Click here to grant required permissions'
        }).props;
        expect(notificationNotice).toBeTruthy();
        await act(() => notificationNotice.onClick());
        expect(mockRequestPermissions).toHaveBeenCalledTimes(1);
    });

    it('user can logout successfull', async () => {
        const tree = create(
            <Provider store={store}>
                <BrowserRouter>
                    <Sidebar />
                </BrowserRouter>
            </Provider>
        );
        let input = tree.root.findByProps({
            id: 'logoutBtn'
        }).props;
        const resetAuthSpyOn = jest.spyOn(auth, 'resetAuthToken');
        await act(() => input.onClick());
        expect(navigate).toHaveBeenCalledWith('/');
        mockApiResponse();
        await store.dispatch(logoutUser({}));
        expect(resetAuthSpyOn).toHaveBeenCalledTimes(1);
    });
});
