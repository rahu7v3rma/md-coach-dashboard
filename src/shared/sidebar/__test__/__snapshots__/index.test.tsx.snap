// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar SideBar snapshot 1`] = `
<div
  className="sc-ckPVkh wbexr"
>
  <div
    className="sc-gAigJI hCiGyo"
  >
    <div
      className="sc-hCnrGf gQrVyO"
    >
      <img
        alt="MD Icon"
        className="sc-jWActB hOKcqo"
        src={Object {}}
      />
      <p
        className="sc-gCkVGe byyJTC"
      >
        Mastering 
        <br />
        Programs
      </p>
    </div>
    <div
      className="sc-iWxGtA eBJcwv"
    >
      <div
        className="sc-pDgPE lkMphx"
        onClick={[Function]}
        selected={false}
      >
        <div
          className="sc-hIiahf gceOdP"
        >
          <img
            alt="My Clients"
            src={Object {}}
          />
        </div>
        <p
          className="sc-IeChK bthVaH"
          selected={false}
        >
          My Clients
        </p>
      </div>
      <div
        className="sc-pDgPE lkMphx"
        onClick={[Function]}
        selected={false}
      >
        <div
          className="sc-hIiahf gceOdP"
        >
          <img
            alt="Chat"
            src={Object {}}
          />
        </div>
        <p
          className="sc-IeChK bthVaH"
          selected={false}
        >
          Chat
        </p>
      </div>
    </div>
    <p
      className="sc-bgqQcB kExYlE sc-bNLxdK bCWNVb"
      color="#B16663"
      fontSize={12}
      fontWeight="500"
      onClick={[Function]}
    >
      Notifications are disabled. Click here to grant required permissions
    </p>
  </div>
  <div
    className="sc-bhlPkD eCkjlf"
  >
    <div
      className="sc-fwEaKe juobiZ"
    >
      <div
        className="sc-hpGEOZ kVoPHL"
      >
        <p
          className="sc-bgqQcB clLeeq"
          fontWeight="600"
        >
          Logout
        </p>
        <p
          className="sc-bgqQcB cGGHIb"
          color="#A4AAAF"
          fontSize={11}
          fontWeight="500"
        >
          Click below to logout the dashboard
        </p>
      </div>
      <button
        className="sc-ispOId humVbF"
        height="40px"
        id="logoutBtn"
        onClick={[Function]}
      >
        <p
          className="sc-bgqQcB cDRZfz"
          fontSize={12}
          fontWeight="600"
        >
          Logout
        </p>
      </button>
    </div>
  </div>
</div>
`;
